homeassistant/components/nasweb/__init__.py

```diff
@@ -0,0 +1,102 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr
+
+from .commons import get_hass_address_from_entry, initialize_notification_coordinator
+from .const import DOMAIN, MANUFACTURER, NASWEB_CONFIG_URL, NOTIFY_COORDINATOR
+from .coordinator import NASwebCoordinator, NotificationCoordinator
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up NASweb from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+    notifi_coordinator = hass.data[DOMAIN].get(NOTIFY_COORDINATOR)
+    if notifi_coordinator is None:
+        notifi_coordinator = initialize_notification_coordinator(hass)
+        if notifi_coordinator is None:
+            _LOGGER.error("Failed to initialize coordinator")
+            return False
+        hass.data[DOMAIN][NOTIFY_COORDINATOR] = notifi_coordinator
+
+    webio_api = WebioAPI(
+        entry.data[CONF_HOST], entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD]
+    )
+    if not await webio_api.check_connection():
+        _LOGGER.error("WebioAPI.check_connection: failed")
+        return False
+    if not await webio_api.refresh_device_info():
+        _LOGGER.error("WebioAPI.refresh_device_info: failed")
+        return False
+    webio_serial = webio_api.get_serial_number()
+    if webio_serial is None:
+        _LOGGER.error("WebIO serial number is not available")
+        return False
+
+    coordinator = NASwebCoordinator(
+        hass, webio_api, name=f"NASweb[{webio_api.get_name()}]"
+    )
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+    notifi_coordinator.add_coordinator(webio_serial, coordinator)
+
+    hass_address = get_hass_address_from_entry(hass, entry.data)
+    if hass_address is None:
+        _LOGGER.error("Cannot determine hass address for webio")
+        return False
+    if not await webio_api.status_subscription(hass_address, True):
+        _LOGGER.error("Failed to subscribe for status updates from webio")
+        return False
+
+    if not await notifi_coordinator.check_connection(webio_serial):
+        _LOGGER.error(
+            "Wasn't able to confirm connection with webio. Check form data and try again"
+        )
+        return False
+
+    device_registry = dr.async_get(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, webio_serial)},
+        manufacturer=MANUFACTURER,
+        name=webio_api.get_name(),
+        configuration_url=NASWEB_CONFIG_URL.replace("[host]", entry.data[CONF_HOST]),
+    )
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```

Comment:
If all config entries are unloaded, we should stop/destroy the coordinator (if that's needed) and clear pop hass.data[DOMAIN]

Reply:
That was my initial approach but i ran into a problem. The hass.http.app.router.add_post() function call seems to assign notification handler for the duration of HA runtime and cannot be removed (I haven't found a way to do this) or overwritten (throws an exception). If I remove coordinator, all future tries to add this integration to HA would fail because assigning new notification handler throws an exception (until HA restart).

Reply:
We don't allow integrations do directly interact with `hass.http.router`.
You should either register a view, or maybe more appropriate, a webhook to allow the API to interact with Home Assistant.